File: website/object_parsing.js
// object_parsing.js
// JavaScript version of the Interactive class using Three.js for 3D object manipulation
// This class handles zoom, rotate, update, show, and reset for a 3D object

import * as THREE from 'https://unpkg.com/three@0.128.0/build/three.module.js';
import { OBJLoader } from 'https://unpkg.com/three@0.128.0/examples/jsm/loaders/OBJLoader.js';

export class Interactive {
  constructor(objPath, scene, onLoad) {
    this.scene = scene;
    this.currentScale = 1.0;
    this.originalCenter = new THREE.Vector3(0, 0, 0);
    this.mesh = null;
    this.ORIGINAL_OBJECT = null;
    this.loader = new OBJLoader();
    this.loader.load(objPath, (object) => {
      // Assume the OBJ is a single mesh
      this.mesh = object.children[0];
      this.mesh.geometry.computeBoundingBox();
      this.originalCenter = this.mesh.geometry.boundingBox.getCenter(new THREE.Vector3());
      this.ORIGINAL_OBJECT = this.mesh.clone();
      this.scene.add(this.mesh);
      if (onLoad) onLoad(this.mesh);
    });
  }

  zoom(pinchSpeed) {
    const maxSpeed = 3.0;
    const minSpeed = -3.0;
    pinchSpeed = Math.max(Math.min(pinchSpeed, maxSpeed), minSpeed);
    const zoomFactor = 1 + pinchSpeed * 0.05;
    const proposedScale = this.currentScale * zoomFactor;
    const minScale = 1.0;
    const maxScale = 5.0;
    const clampedScale = Math.max(Math.min(proposedScale, maxScale), minScale);
    const scaleChange = clampedScale / this.currentScale;
    if (this.mesh) {
      this.mesh.scale.multiplyScalar(scaleChange);
    }
    this.currentScale = clampedScale;
  }

  rotate(rotationAxis, angleRad) {
    if (!this.mesh) return;
    const axis = new THREE.Vector3(...rotationAxis);
    if (axis.length() === 0) return;
    axis.normalize();
    // Rotate around the object's center
    const center = this.originalCenter.clone();
    this.mesh.position.sub(center);
    this.mesh.applyMatrix4(new THREE.Matrix4().makeRotationAxis(axis, angleRad));
    this.mesh.position.add(center);
    this.mesh.geometry.computeVertexNormals();
  }

  update({ zoomFactor = 1.0, rotationAngle = 0.0, rotationAxis = [0, 0, 0] }) {
    if (zoomFactor !== 1.0) {
      this.zoom(zoomFactor);
    }
    if (rotationAngle !== 0.0 || rotationAxis.some(v => v !== 0)) {
      this.rotate(rotationAxis, rotationAngle);
    }
  }

  show() {
    // In Three.js, the mesh is already in the scene and rendered by the animation loop
    // This method can be used to trigger a render if needed
  }

  reset() {
    if (!this.mesh || !this.ORIGINAL_OBJECT) return;
    // Remove current mesh
    this.scene.remove(this.mesh);
    // Clone the original object
    this.mesh = this.ORIGINAL_OBJECT.clone();
    this.currentScale = 1.0;
    this.scene.add(this.mesh);
  }
} 

File: original_python_code/object_parsing.py
import os
import numpy as np
import trimesh
import open3d as o3d
import copy

class Interactive():

  def __init__(self, obj_path):
    self.trimesh_obj = trimesh.load(obj_path, force = 'mesh')

    self.mesh = o3d.geometry.TriangleMesh(
      vertices = o3d.utility.Vector3dVector(self.trimesh_obj.vertices),
      triangles = o3d.utility.Vector3iVector(self.trimesh_obj.faces)
    )
    self.mesh.compute_vertex_normals()

    if hasattr(self.trimesh_obj.visual, 'vertex_colors') and self.trimesh_obj.visual.vertex_colors is not None:
      colors = self.trimesh_obj.visual.vertex_colors[:, :3] / 255.0
      self.mesh.vertex_colors = o3d.utility.Vector3dVector(colors)

    self.original_center = self.mesh.get_center()
    self.ORIGINAL_OBJECT = copy.deepcopy(self.mesh)
    self.current_scale = 1.0

  def zoom(self, pinch_speed):
    max_speed = 3.0
    min_speed = -3.0
    pinch_speed = max(min(pinch_speed, max_speed), min_speed)

    zoom_factor = 1 + pinch_speed * 0.05
    proposed_scale = self.current_scale * zoom_factor

    min_scale = 1.0
    max_scale = 5.0

    clamped_scale = max(min(proposed_scale, max_scale), min_scale)
    zoom_factor = clamped_scale / self.current_scale

    self.mesh.scale(zoom_factor, center=self.original_center)
    self.current_scale = clamped_scale


  def rotate(self, rotation_axis: list, angle_rad: float):
    axis = np.array(rotation_axis, dtype=float)
    axis /= np.linalg.norm(axis)

    rotation_matrix = trimesh.transformations.rotation_matrix(
      angle_rad, axis, point=self.original_center
    )

    vertices = np.asarray(self.mesh.vertices)
    num_vertices = vertices.shape[0]
    vertices_homogeneous = np.hstack((vertices, np.ones((num_vertices, 1))))
    transformed_vertices = (rotation_matrix @ vertices_homogeneous.T).T[:, :3]

    self.mesh.vertices = o3d.utility.Vector3dVector(transformed_vertices)
    self.mesh.compute_vertex_normals()

  def update(self, zoom_factor = 1.0, rotation_angle = 0.0, rotation_axis = [0, 0, 0]):
    if zoom_factor != 1.0:
      self.zoom(zoom_factor)
    if rotation_angle != 0.0 or rotation_axis != [0, 0, 0]:
      self.rotate(rotation_axis, rotation_angle)

  def show(self):
    o3d.visualization.draw_geometries([self.mesh])

  def reset(self):
    self.mesh = o3d.geometry.TriangleMesh(
      vertices=o3d.utility.Vector3dVector(self.ORIGINAL_OBJECT.vertices),
      triangles=o3d.utility.Vector3iVector(self.ORIGINAL_OBJECT.faces)
    )
    self.mesh.compute_vertex_normals()